File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/AlignReactiveScopesToBlockScopesHIR.ts
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {CompilerError} from '..';
import {
  BlockId,
  HIRFunction,
  InstructionId,
  MutableRange,
  Place,
  ReactiveScope,
  makeInstructionId,
} from '../HIR/HIR';
import {
  eachInstructionLValue,
  eachInstructionValueOperand,
  eachTerminalOperand,
  mapTerminalSuccessors,
  terminalFallthrough,
} from '../HIR/visitors';
import {retainWhere_Set} from '../Utils/utils';
import {getPlaceScope} from './BuildReactiveBlocks';

type InstructionRange = MutableRange;
/*
 * Note: this is the 2nd of 4 passes that determine how to break a function into discrete
 * reactive scopes (independently memoizeable units of code):
 * 1. InferReactiveScopeVariables (on HIR) determines operands that mutate together and assigns
 *     them a unique reactive scope.
 * 2. AlignReactiveScopesToBlockScopes (this pass, on ReactiveFunction) aligns reactive scopes
 *     to block scopes.
 * 3. MergeOverlappingReactiveScopes (on ReactiveFunction) ensures that reactive scopes do not
 *     overlap, merging any such scopes.
 * 4. BuildReactiveBlocks (on ReactiveFunction) groups the statements for each scope into
 *     a ReactiveScopeBlock.
 *
 * Prior inference passes assign a reactive scope to each operand, but the ranges of these
 * scopes are based on specific instructions at arbitrary points in the control-flow graph.
 * However, to codegen blocks around the instructions in each scope, the scopes must be
 * aligned to block-scope boundaries - we can't memoize half of a loop!
 *
 * This pass updates reactive scope boundaries to align to control flow boundaries, for
 * example:
 *
 * ```javascript
 * function foo(cond, a) {
 *                     ⌵ original scope
 *                          ⌵ expanded scope
 *    const x = [];    ⌝    ⌝
 *    if (cond) {      ⎮    ⎮
 *      ...            ⎮    ⎮
 *      x.push(a);     ⌟    ⎮
 *      ...                 ⎮
 *    }                     ⌟
 * }
 * ```
 *
 * Here the original scope for `x` ended partway through the if consequent, but we can't
 * memoize part of that block. This pass would align the scope to the end of the consequent.
 *
 * The more general rule is that a reactive scope may only end at the same block scope as it
 * began: this pass therefore finds, for each scope, the block where that scope started and
 * finds the first instruction after the scope's mutable range in that same block scope (which
 * will be the updated end for that scope).
 */
export function alignReactiveScopesToBlockScopesHIR(fn: HIRFunction): void {
  const activeBlockFallthroughRanges: Array<{
    range: InstructionRange;
    fallthrough: BlockId;
  }> = [];
  const activeScopes = new Set<ReactiveScope>();
  const seen = new Set<ReactiveScope>();
  const valueBlockNodes = new Map<BlockId, ValueBlockNode>();
  const placeScopes = new Map<Place, ReactiveScope>();

  function recordPlace(
    id: InstructionId,
    place: Place,
    node: ValueBlockNode | null,
  ): void {
    if (place.identifier.scope !== null) {
      placeScopes.set(place, place.identifier.scope);
    }

    const scope = getPlaceScope(id, place);
    if (scope == null) {
      return;
    }
    activeScopes.add(scope);
    node?.children.push({kind: 'scope', scope, id});

    if (seen.has(scope)) {
      return;
    }
    seen.add(scope);
    if (node != null && node.valueRange !== null) {
      scope.range.start = makeInstructionId(
        Math.min(node.valueRange.start, scope.range.start),
      );
      scope.range.end = makeInstructionId(
        Math.max(node.valueRange.end, scope.range.end),
      );
    }
  }

  for (const [, block] of fn.body.blocks) {
    const startingId = block.instructions[0]?.id ?? block.terminal.id;
    retainWhere_Set(activeScopes, scope => scope.range.end > startingId);
    const top = activeBlockFallthroughRanges.at(-1);
    if (top?.fallthrough === block.id) {
      activeBlockFallthroughRanges.pop();
      /*
       * All active scopes must have either started before or within the last
       * block-fallthrough range. In either case, they overlap this block-
       * fallthrough range and can have their ranges extended.
       */
      for (const scope of activeScopes) {
        scope.range.start = makeInstructionId(
          Math.min(scope.range.start, top.range.start),
        );
      }
    }

    const {instructions, terminal} = block;
    const node = valueBlockNodes.get(block.id) ?? null;
    for (const instr of instructions) {
      for (const lvalue of eachInstructionLValue(instr)) {
        recordPlace(instr.id, lvalue, node);
      }
      for (const operand of eachInstructionValueOperand(instr.value)) {
        recordPlace(instr.id, operand, node);
      }
    }
    for (const operand of eachTerminalOperand(terminal)) {
      recordPlace(terminal.id, operand, node);
    }

    const fallthrough = terminalFallthrough(terminal);
    if (fallthrough !== null) {
      /*
       * Any currently active scopes that overlaps the block-fallthrough range
       * need their range extended to at least the first instruction of the
       * fallthrough
       */
      const fallthroughBlock = fn.body.blocks.get(fallthrough)!;
      const nextId =
        fallthroughBlock.instructions[0]?.id ?? fallthroughBlock.terminal.id;
      for (const scope of activeScopes) {
        if (scope.range.end > terminal.id) {
          scope.range.end = makeInstructionId(
            Math.max(scope.range.end, nextId),
          );
        }
      }
      /**
       * We also record the block-fallthrough range for future scopes that begin
       * within the range (and overlap with the range end).
       */
      activeBlockFallthroughRanges.push({
        fallthrough,
        range: {
          start: terminal.id,
          end: nextId,
        },
      });

      CompilerError.invariant(!valueBlockNodes.has(fallthrough), {
        reason: 'Expect hir blocks to have unique fallthroughs',
        loc: terminal.loc,
      });
      if (node != null) {
        valueBlockNodes.set(fallthrough, node);
      }
    }

    /*
     * Visit all successors (not just direct successors for control-flow ordering) to
     * set a value block node where necessary to align the value block start/end
     * back to the outer block scope.
     *
     * TODO: add a variant of eachTerminalSuccessor() that visits _all_ successors, not
     * just those that are direct successors for normal control-flow ordering.
     */
    mapTerminalSuccessors(terminal, successor => {
      if (valueBlockNodes.has(successor)) {
        return successor;
      }

      const successorBlock = fn.body.blocks.get(successor)!;
      if (successorBlock.kind === 'block' || successorBlock.kind === 'catch') {
        /*
         * we need the block kind check here because the do..while terminal's
         * successor is a block, and try's successor is a catch block
         */
      } else if (
        node == null ||
        terminal.kind === 'ternary' ||
        terminal.kind === 'logical' ||
        terminal.kind === 'optional'
      ) {
        /**
         * Create a new node whenever we transition from non-value -> value block.
         *
         * For compatibility with the previous ReactiveFunction-based scope merging logic,
         * we also create new scope nodes for ternary, logical, and optional terminals.
         * Inside value blocks we always store a range (valueRange) that is the
         * start/end instruction ids at the nearest parent block scope level, so that
         * scopes inside the value blocks can be extended to align with block scope
         * instructions.
         */
        let valueRange: MutableRange;
        if (node == null) {
          // Transition from block->value block, derive the outer block range
          CompilerError.invariant(fallthrough !== null, {
            reason: `Expected a fallthrough for value block`,
            loc: terminal.loc,
          });
          const fallthroughBlock = fn.body.blocks.get(fallthrough)!;
          const nextId =
            fallthroughBlock.instructions[0]?.id ??
            fallthroughBlock.terminal.id;
          valueRange = {
            start: terminal.id,
            end: nextId,
          };
        } else {
          // else value->value transition, reuse the range
          valueRange = node.valueRange;
        }
        const childNode: ValueBlockNode = {
          kind: 'node',
          id: terminal.id,
          children: [],
          valueRange,
        };
        node?.children.push(childNode);
        valueBlockNodes.set(successor, childNode);
      } else {
        // this is a value -> value block transition, reuse the node
        valueBlockNodes.set(successor, node);
      }
      return successor;
    });
  }
}

type ValueBlockNode = {
  kind: 'node';
  id: InstructionId;
  valueRange: MutableRange;
  children: Array<ValueBlockNode | ReactiveScopeNode>;
};
type ReactiveScopeNode = {
  kind: 'scope';
  id: InstructionId;
  scope: ReactiveScope;
};

function _debug(node: ValueBlockNode): string {
  const buf: Array<string> = [];
  _printNode(node, buf, 0);
  return buf.join('\n');
}
function _printNode(
  node: ValueBlockNode | ReactiveScopeNode,
  out: Array<string>,
  depth: number = 0,
): void {
  let prefix = '  '.repeat(depth);
  if (node.kind === 'scope') {
    out.push(
      `${prefix}[${node.id}] @${node.scope.id} [${node.scope.range.start}:${node.scope.range.end}]`,
    );
  } else {
    let range = ` (range=[${node.valueRange.start}:${node.valueRange.end}])`;
    out.push(`${prefix}[${node.id}] node${range} [`);
    for (const child of node.children) {
      _printNode(child, out, depth + 1);
    }
    out.push(`${prefix}]`);
  }
}

File: compiler/crates/react_compiler_inference/src/align_reactive_scopes_to_block_scopes_hir.rs
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

//! Aligns reactive scope boundaries to block scope boundaries in the HIR.
//!
//! Ported from TypeScript `src/ReactiveScopes/AlignReactiveScopesToBlockScopesHIR.ts`.
//!
//! This is the 2nd of 4 passes that determine how to break a function into
//! discrete reactive scopes (independently memoizable units of code):
//! 1. InferReactiveScopeVariables (on HIR) determines operands that mutate
//!    together and assigns them a unique reactive scope.
//! 2. AlignReactiveScopesToBlockScopes (this pass) aligns reactive scopes
//!    to block scopes.
//! 3. MergeOverlappingReactiveScopes ensures scopes do not overlap.
//! 4. BuildReactiveBlocks groups the statements for each scope.
//!
//! Prior inference passes assign a reactive scope to each operand, but the
//! ranges of these scopes are based on specific instructions at arbitrary
//! points in the control-flow graph. However, to codegen blocks around the
//! instructions in each scope, the scopes must be aligned to block-scope
//! boundaries — we can't memoize half of a loop!

use std::collections::{HashMap, HashSet};

use react_compiler_hir::environment::Environment;
use react_compiler_hir::visitors;
use react_compiler_hir::visitors::{
    each_instruction_lvalue_ids, each_instruction_value_operand_ids, each_terminal_operand_ids,
};
use react_compiler_hir::{
    BlockId, BlockKind, EvaluationOrder, HirFunction, IdentifierId,
    MutableRange, ScopeId, Terminal,
};

// =============================================================================
// ValueBlockNode — stores the valueRange for scope alignment in value blocks
// =============================================================================

/// Tracks the value range for a value block. The `children` field from the TS
/// implementation is only used for debug output and is omitted here.
#[derive(Clone)]
struct ValueBlockNode {
    value_range: MutableRange,
}

/// Returns all block IDs referenced by a terminal, including both direct
/// successors and fallthrough.
fn all_terminal_block_ids(terminal: &Terminal) -> Vec<BlockId> {
    visitors::each_terminal_all_successors(terminal)
}

// =============================================================================
// Helper: get the first EvaluationOrder in a block
// =============================================================================

fn block_first_id(func: &HirFunction, block_id: BlockId) -> EvaluationOrder {
    let block = func.body.blocks.get(&block_id).unwrap();
    if !block.instructions.is_empty() {
        func.instructions[block.instructions[0].0 as usize].id
    } else {
        block.terminal.evaluation_order()
    }
}

// =============================================================================
// BlockFallthroughRange
// =============================================================================

#[derive(Clone)]
struct BlockFallthroughRange {
    fallthrough: BlockId,
    range: MutableRange,
}

// =============================================================================
// Public API
// =============================================================================

/// Aligns reactive scope boundaries to block scope boundaries in the HIR.
///
/// This pass updates reactive scope boundaries to align to control flow
/// boundaries. For example, if a scope ends partway through an if consequent,
/// the scope is extended to the end of the consequent block.
pub fn align_reactive_scopes_to_block_scopes_hir(func: &mut HirFunction, env: &mut Environment) {
    let mut active_block_fallthrough_ranges: Vec<BlockFallthroughRange> = Vec::new();
    let mut active_scopes: HashSet<ScopeId> = HashSet::new();
    let mut seen: HashSet<ScopeId> = HashSet::new();
    let mut value_block_nodes: HashMap<BlockId, ValueBlockNode> = HashMap::new();

    let block_ids: Vec<BlockId> = func.body.blocks.keys().copied().collect();

    for &block_id in &block_ids {
        let starting_id = block_first_id(func, block_id);

        // Retain only active scopes whose range.end > startingId
        active_scopes.retain(|&scope_id| {
            env.scopes[scope_id.0 as usize].range.end > starting_id
        });

        // Check if we've reached a fallthrough block
        if let Some(top) = active_block_fallthrough_ranges.last().cloned() {
            if top.fallthrough == block_id {
                active_block_fallthrough_ranges.pop();
                // All active scopes overlap this block-fallthrough range;
                // extend their start to include the range start.
                for &scope_id in &active_scopes {
                    let scope = &mut env.scopes[scope_id.0 as usize];
                    scope.range.start = std::cmp::min(scope.range.start, top.range.start);
                }
            }
        }

        let node = value_block_nodes.get(&block_id).cloned();

        // Visit instruction lvalues and operands
        let block = func.body.blocks.get(&block_id).unwrap();
        let instr_ids: Vec<react_compiler_hir::InstructionId> =
            block.instructions.iter().copied().collect();
        for &instr_id in &instr_ids {
            let instr = &func.instructions[instr_id.0 as usize];
            let eval_order = instr.id;

            let lvalue_ids = each_instruction_lvalue_ids(instr);
            for lvalue_id in lvalue_ids {
                record_place_id(
                    eval_order,
                    lvalue_id,
                    &node,
                    env,
                    &mut active_scopes,
                    &mut seen,
                );
            }

            let operand_ids = each_instruction_value_operand_ids(&instr.value, env);
            for operand_id in operand_ids {
                record_place_id(
                    eval_order,
                    operand_id,
                    &node,
                    env,
                    &mut active_scopes,
                    &mut seen,
                );
            }
        }

        // Visit terminal operands
        let block = func.body.blocks.get(&block_id).unwrap();
        let terminal_eval_order = block.terminal.evaluation_order();
        let terminal_operand_ids = each_terminal_operand_ids(&block.terminal);
        for operand_id in terminal_operand_ids {
            record_place_id(
                terminal_eval_order,
                operand_id,
                &node,
                env,
                &mut active_scopes,
                &mut seen,
            );
        }

        let block = func.body.blocks.get(&block_id).unwrap();
        let terminal = &block.terminal;
        let fallthrough = visitors::terminal_fallthrough(terminal);
        let is_branch = matches!(terminal, Terminal::Branch { .. });
        let is_goto = match terminal {
            Terminal::Goto { block, .. } => Some(*block),
            _ => None,
        };
        let is_ternary_logical_optional = matches!(
            terminal,
            Terminal::Ternary { .. } | Terminal::Logical { .. } | Terminal::Optional { .. }
        );
        let all_successors = all_terminal_block_ids(terminal);

        // Handle fallthrough logic
        if let Some(ft) = fallthrough {
            if !is_branch {
                let next_id = block_first_id(func, ft);

                for &scope_id in &active_scopes {
                    let scope = &mut env.scopes[scope_id.0 as usize];
                    if scope.range.end > terminal_eval_order {
                        scope.range.end = std::cmp::max(scope.range.end, next_id);
                    }
                }

                active_block_fallthrough_ranges.push(BlockFallthroughRange {
                    fallthrough: ft,
                    range: MutableRange {
                        start: terminal_eval_order,
                        end: next_id,
                    },
                });

                assert!(
                    !value_block_nodes.contains_key(&ft),
                    "Expect hir blocks to have unique fallthroughs"
                );
                if let Some(n) = &node {
                    value_block_nodes.insert(ft, n.clone());
                }
            }
        } else if let Some(goto_block) = is_goto {
            // Handle goto to label
            let start_pos = active_block_fallthrough_ranges
                .iter()
                .position(|r| r.fallthrough == goto_block);
            let top_idx = if active_block_fallthrough_ranges.is_empty() {
                None
            } else {
                Some(active_block_fallthrough_ranges.len() - 1)
            };
            if let Some(pos) = start_pos {
                if top_idx != Some(pos) {
                    let start_range = active_block_fallthrough_ranges[pos].clone();
                    let first_id = block_first_id(func, start_range.fallthrough);

                    for &scope_id in &active_scopes {
                        let scope = &mut env.scopes[scope_id.0 as usize];
                        if scope.range.end <= terminal_eval_order {
                            continue;
                        }
                        scope.range.start =
                            std::cmp::min(start_range.range.start, scope.range.start);
                        scope.range.end = std::cmp::max(first_id, scope.range.end);
                    }
                }
            }
        }

        // Visit all successors to set up value block nodes
        for successor in all_successors {
            if value_block_nodes.contains_key(&successor) {
                continue;
            }

            let successor_block = func.body.blocks.get(&successor).unwrap();
            if successor_block.kind == BlockKind::Block
                || successor_block.kind == BlockKind::Catch
            {
                // Block or catch kind: don't create a value block node
            } else if node.is_none() || is_ternary_logical_optional {
                // Create a new node when transitioning non-value -> value,
                // or for ternary/logical/optional terminals.
                let value_range = if node.is_none() {
                    // Transition from block -> value block
                    let ft =
                        fallthrough.expect("Expected a fallthrough for value block");
                    let next_id = block_first_id(func, ft);
                    MutableRange {
                        start: terminal_eval_order,
                        end: next_id,
                    }
                } else {
                    // Value -> value transition (ternary/logical/optional): reuse range
                    node.as_ref().unwrap().value_range.clone()
                };

                value_block_nodes.insert(
                    successor,
                    ValueBlockNode { value_range },
                );
            } else {
                // Value -> value block transition: reuse the node
                if let Some(n) = &node {
                    value_block_nodes.insert(successor, n.clone());
                }
            }
        }
    }

    // Sync identifier mutable_range with their scope's range.
    // In TS, identifier.mutableRange and scope.range are the same shared object,
    // so modifications to scope.range are automatically visible through the
    // identifier. In Rust they are separate copies, so we must explicitly sync.
    for ident in &mut env.identifiers {
        if let Some(scope_id) = ident.scope {
            let scope_range = &env.scopes[scope_id.0 as usize].range;
            ident.mutable_range.start = scope_range.start;
            ident.mutable_range.end = scope_range.end;
        }
    }
}

/// Records a place's scope as active and adjusts scope ranges for value blocks.
///
/// Mirrors TS `recordPlace(id, place, node)`.
fn record_place_id(
    id: EvaluationOrder,
    identifier_id: IdentifierId,
    node: &Option<ValueBlockNode>,
    env: &mut Environment,
    active_scopes: &mut HashSet<ScopeId>,
    seen: &mut HashSet<ScopeId>,
) {
    // Get the scope for this identifier, if active at this instruction
    let scope_id = match env.identifiers[identifier_id.0 as usize].scope {
        Some(scope_id) => {
            let scope = &env.scopes[scope_id.0 as usize];
            if id >= scope.range.start && id < scope.range.end {
                Some(scope_id)
            } else {
                None
            }
        }
        None => None,
    };

    if let Some(scope_id) = scope_id {
        active_scopes.insert(scope_id);

        if seen.contains(&scope_id) {
            return;
        }
        seen.insert(scope_id);

        if let Some(n) = node {
            let scope = &mut env.scopes[scope_id.0 as usize];
            scope.range.start = std::cmp::min(n.value_range.start, scope.range.start);
            scope.range.end = std::cmp::max(n.value_range.end, scope.range.end);
        }
    }
}
